pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/RegexpExtractTransformFunction.java

```diff
@@ -0,0 +1,119 @@
+ * the regexp in `value`, otherwise it returns the first match.
+ */
+public class RegexpExtractTransformFunction extends BaseTransformFunction {
+  public static final String FUNCTION_NAME = "REGEXP_EXTRACT";
+
+  private TransformFunction _valueFunction;
+  private Pattern _regexp;
+  private int _position;
+  private int _occurrence;
+  private String[] _stringOutputRegexMatches;
+  private TransformResultMetadata _resultMetadata;
+
+  @Override
+  public String getName() {
+    return FUNCTION_NAME;
+  }
+
+  @Override
+  public void init(List<TransformFunction> arguments, Map<String, DataSource> dataSourceMap) {
+    Preconditions.checkArgument(
+        arguments.size() >= 2 && arguments.size() <= 4,
+        "REGEXP_EXTRACT takes between 2 to 4 arguments. See usage: "
+            + "REGEXP_EXTRACT(`value`, `regexp`[, `pos`, `occurrence`]");
+    _valueFunction = arguments.get(0);
+
+    TransformFunction regexpFunction = arguments.get(1);
+    Preconditions.checkState(
+        regexpFunction instanceof LiteralTransformFunction,
+        "`regexp` must be a literal regex expression.");
+    _regexp = Pattern.compile(((LiteralTransformFunction) regexpFunction).getLiteral());
+
+    if (arguments.size() >= 3) {
+      TransformFunction positionFunction = arguments.get(2);
+      Preconditions.checkState(positionFunction instanceof LiteralTransformFunction
+              && Integer.parseInt(((LiteralTransformFunction) positionFunction).getLiteral()) > 0,
+          "`pos` must be a literal, positive number.");
+      _position = Integer.parseInt(((LiteralTransformFunction) positionFunction).getLiteral());
+    } else {
+      _position = 1;
+    }
+    if (arguments.size() == 4) {
+      TransformFunction occurrenceFunction = arguments.get(3);
+      Preconditions.checkState(occurrenceFunction instanceof LiteralTransformFunction
+              && Integer.parseInt(((LiteralTransformFunction) occurrenceFunction).getLiteral()) > 0,
+          "`pos` must be a literal, positive number.");
+      _occurrence = Integer.parseInt(((LiteralTransformFunction) occurrenceFunction).getLiteral());
+    } else {
+      _occurrence = 1;
+    }
+    _resultMetadata = STRING_SV_NO_DICTIONARY_METADATA;
+    _stringOutputRegexMatches = new String[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+  }
+
+  @Override
+  public TransformResultMetadata getResultMetadata() {
+    return _resultMetadata;
+  }
+
+  @Override
+  public String[] transformToStringValuesSV(ProjectionBlock projectionBlock) {
+    int length = projectionBlock.getNumDocs();
+    String[] valuesSV = _valueFunction.transformToStringValuesSV(projectionBlock);
+    for (int i = 0; i < length; ++i) {
+      Matcher matcher = _regexp.matcher(valuesSV[i].substring(_position - 1));
+      if (matcher.find()) {
+        _stringOutputRegexMatches[i] = matcher.group(_occurrence - 1);
+      } else {
+        _stringOutputRegexMatches[i] = null;
```

Comment:
We cannot put `null` in the response. Need to pick a default null value, or make it configurable from the function argument

Reply:
> Should we consider following the presto semantic of regexp_extract: https://prestodb.io/docs/current/functions/regexp.html
> Seems databricks is also using this semantic: https://docs.databricks.com/sql/language-manual/functions/regexp_extract.html

if i understand correctly this only means removing support for position right? (e.g. move occurrence to 3rd argument)

Reply:
E.g. put an optional 4th argument for the default, or use `"null"` or empty string if not specified

Reply:
oh. if we can use empty string or `"null"` then its super easy. I was worried that we need to infer the default based on context. which could be hard. will implement this